docs/proposals/dns-using-envoy.md

```diff
@@ -0,0 +1,73 @@
+# Use Envoy DNSFilter for DNS resolution
+
+## Context
+
+Currently, Kuma relies on an external DNS to provide VIPs which are then added as services which proxy back to the actual Cluster in Envoy.
+For VIPs to be resolved Kuma CP runs a DNS server which maintains this mapping.
+
+This current setup has some limitations:
+
+- Apps need to set their DNS to point to the CP's DNS (or CNIs need to configure DNS to be correct)
+- The CP needs to run a DNS and handle the load
+- When the CP is down DNS requests will fail. This is usually not the case in a Service Mesh
+- DNS configuration is global to all meshes
+- DNS is a separate component which makes it less customisable compared to things that end up in XDS
+- This DNS is not aware of the origin of requests, therefore isolation of meshes breaks as anyone can resolve any domain
+
+In this proposal we suggest using [Envoy's DNSFilter](https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/udp_filters/dns_filter) to replace the CP DNS.
+We add a new `DNS policy` to be able to control the configuration of the DNS.
+
+## Out of scope
+
+* UDP transparent proxying
+* Handling SRV records. While DNSFilter supports it, it's probably not necessary in v1
+
+## Requirements
+
+* The top-level domain should be configurable (defaults to `mesh`)
+* Drop in replacement to Kuma DNS (Default configuration should match Kuma's current configuration)
+* ability to set upstream dns resolvers to have a single resolver in the configuration
```

Comment:
How is this going to work in the Universal deployments? Today with Transparent Proxying, we do edit /etc/resolv.conf, which is tricky, yet we preserve the original servers.

Reply:
I think you'd do the same (update the /etc/resolv.conf).

For example in our universal deployment we set an iptable rule to direct to whatever address you'd like .e.g: `-A CNI-o-631b6b8422a8498934cee4 -d 192.168.0.53/32 -p udp -m udp --dport 53 -j REDIRECT --to-ports <passthrough-outbound-port>` and the /etc/resolv.conf to this as well.